### PR TITLE
"Map tools" for the Map Editor, cross-mods

### PR DESCRIPTION
**Pull request by FRenzy-OpenRA**
(First PR / piece of OpenRA code for me :) )

Adds a new set of tools for the Map Editor. 
Interface : a new button "Map tools" into the map editor menu. 

The tools currently include : Append, Copy, Mirror (Left or Top), Resize.
There is an option to choose what is affected by the operation :Tiles, Resources, Actors

Support added for all mods (rectangular and isometric maps). Only the Resize for isometric maps gives scrambled results, due to the nature of iso grid.

Some screenshots in the forum post : https://forum.openra.net/viewtopic.php?f=83&t=20568&p=307886#p307886

----------------------
Ideas for future improvements (out of scope for now) : 
- more tools (e.g. : "Rotate" for antisymmetry)
- improving "Resize" for iso maps
- polished interface
- map selector
- in-editor local tools ? e.g. : mirroring a selection, ...

### CODE REVIEW

**Comment by GraionDilach** [comment]
Gotta look into this in depth on Sunday, I guess. The stuff you shown on the forums were promising and I guess we can let it slip if the issues with TS won't get fixed in the first run.

**Comment by pchote** [comment]
My main potential concern here is that we have many precedents for things like these to be exposed as utility commands instead of ingame.  I will need to test the PR and/or review the code before I can comment properly, however.

**Comment by GraionDilach** [comment]
You have a point there with it as well.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 1]
```diff
@@ -887,22 +888,14 @@
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
   <Target Name="AfterBuild">
-    <MakeDir Directories="$(SolutionDir)mods/common/"/>
-    <Copy
-      SourceFiles="$(TargetPath)"
-      DestinationFolder="$(SolutionDir)mods/common/"/>
+    <MakeDir Directories="$(SolutionDir)mods/common/" />
```
[the remark this answers is not in the document]
Comment (FRenzy-OpenRA) [reply OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Ok thanks. 
So I remove **all** the changes in OpenRA.Mods.Common.csproj ?

**Comment by FRenzy-OpenRA** [comment]
Thank you both for your quick reply and for your input

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 1 above)
[the remark this answers is not in the document]
Comment (abcdefg30) [reply OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: No, the Compile Include line is ok.

**Comment by abcdefg30** [comment]
`MapsTooltipLogic.cs` is missing the copyright header.

**Comment by netnazgul** [comment]
> My main potential concern here is that we have many precedents for things like these to be exposed as utility commands instead of ingame. 

As I've stated the fact about SirCake's bitmap converter, having map-modifying tools outside the client makes working with maps very tedious and cumbersome because you need to restart the client every time you do a change on the maps on your harddrive (the client only refreshes the map list on launch). This might be alleviated by introducing a "refresh" button for maps somewhere in the client or by making the map editor a separate tool as well.

**Comment by CatGirls420-NerevarII** [comment]
Not all heroes wear capes.

**Comment by pchote** [comment]
Can you please look into porting these to utility commands?  The existing `--resize-map` command (ResizeMapCommand.cs) can be used as a guideline for this.

**Comment by MustaphaTR** [comment]
I belive these, even the resize command, belong to ingame map editor, not utility. While it is far from getting that point, map editor should be self sufficent, at least till you get to do lua/custom.yaml. As netnazgul said, using multipile prohrams for map editing, while one could do, is unnecessarily tedious.

**Comment by FRenzy-OpenRA** [comment]
Thank you both for your feedback.
I've unfortunately run out of time this previous period, and also my objective of learning coding has been put to secondary due to new objectives in life ... 

However whenever I'll get some time in the near future, I'll start to look into the utility implementation of it as a starter. I also want to see if I can make it mod-friendly, and not available only for the official mods. 

When it works all and well, UI integration could be a further step.

**Comment by pchote** [comment]
I'll move this to the future milestone in the meantime.  Feel free to pop in to the IRC channel if you need advice once you are able to get back to this!

**Pull request closed without merging**